core/domain/voiceover_services.py

```diff
@@ -0,0 +1,291 @@
+        voiceover_application)
+
+    voiceover_application_model.put()
+
+
+def get_voiceover_application_by_id(voiceover_application_id):
+    """Returns voiceover application model corresponding to give id.
+
+    Args:
+        voiceover_application_id: str. The voiceover application id.
+
+    Returns:
+        BaseVoiceoverApplication. The voiceover application object for the give
+        application id.
+    """
+    voiceover_application_model = (
+        suggestion_models.GeneralVoiceoverApplicationModel.get_by_id(
+            voiceover_application_id))
+    return _get_voiceover_application_from_model(voiceover_application_model)
+
+
+def get_reviewable_voiceover_applications(user_id):
+    """Returns a list of voiceover applications which the given user can review.
+
+    Args:
+        user_id: str. The user ID of the reviewer.
+
+    Returns:
+        list(BaseVoiceoverApplication). A list of voiceover application which
+        the given user can review.
+    """
+    voiceover_application_models = (
+        suggestion_models.GeneralVoiceoverApplicationModel
+        .get_reviewable_voiceover_applications(user_id))
+
+    return [
+        _get_voiceover_application_from_model(model) for model in (
+            voiceover_application_models)]
+
+
+def get_user_submitted_voiceover_applications(user_id, status=None):
+    """Returns a list of voiceover application submitted by the given user which
+    are currently in the given status.
+
+    Args:
+        user_id: str. The id of the user.
+        status: str|None. The status of the voiceover application.
+
+    Returns:
+        BaseVoiceoverApplication). A list of voiceover application which are
+        submitted by the given user.
+    """
+    voiceover_application_models = (
+        suggestion_models.GeneralVoiceoverApplicationModel
+        .get_user_voiceover_applications(user_id, status))
+
+    return [
+        _get_voiceover_application_from_model(model) for model in (
+            voiceover_application_models)]
+
+
+def accept_voiceover_application(voiceover_application_id, reviewer_id):
+    """Accept the voiceover application of given voiceover application id.
+
+    Args:
+        voiceover_application_id: str. The is of the voiceover application which
+            need to be accepted.
+        reviewer_id: str. The user ID of the reviewer.
+    """
+    voiceover_application = get_voiceover_application_by_id(
+        voiceover_application_id)
+    if reviewer_id == voiceover_application.author_id:
+        raise Exception(
+            'Applicants are not allowed to review their own '
+            'voiceover application.')
+
+    reviewer = user_services.UserActionsInfo(user_id=reviewer_id)
+
+    voiceover_application.accept(reviewer_id)
+
+    _save_voiceover_application(voiceover_application)
+
+
+    if voiceover_application.target_type == feconf.ENTITY_TYPE_EXPLORATION:
+        rights_manager.assign_role_for_exploration(
```

Comment:
Should I move the notification/email and role assignment code in the object `accept` method? (that way I don't have to do if-else here)

Reply:
I think @nithusha21 should advise on this since he's more familiar with the patterns in this part of the codebase. But @DubeySandeep I have one question -- what are the cons of moving it to `accept`? (You've mentioned the pros, but I assume you're asking the question because there are concerns you have as well.)

Reply:
> there are concerns you have as well.

The cons:
We will be going to mix the service layer in the domain layer. (Although we do this in suggestion structure, but I don't think we should make domain object change datastore models. Considering we pass the domain object as a parameter or return type for the functions in service which controller layer uses, so service layer will provide some data to controller layer through which the controller layer can independently change data models.)

@seanlip, any thoughts?

Reply:
Domain objects shouldn't change datastore models. But it seems to me that this PR should follow existing conventions and we should file an issue and later do a separate PR to change everything to fulfill that statement, rather than having two conventions in the codebase simultaneously.

So, I would say: follow existing conventions for similar things currently in the codebase, and change all those things to follow the correct broader convention. These two steps can be done together or separately.

Reply:
> rather than having two conventions in the codebase simultaneously.

I think currently we have this structure!

> These two steps can be done together or separately.

I think we should do it separately, as I've some deliverables which needed to be done quickly! I'll wait for @nithusha21's comment on whether we should leave it as is (if we are going to have this pattern in the future!) or should I move this to the domain class.

(I'm fine with doing either.)

Reply:
> I think currently we have this structure!

Domain objects do not change datastore objects even in suggestion_registry. We use other service layer functions for validating the suggestion. We only read from the datastore. 

> Should I move the notification/email and role assignment code in the object accept method? (that way I don't have to do if-else here)

I don't think so. With feedback threads, we had the service layer handle all the email related tasks. I think we can follow that here too!

Reply:
> Domain objects do not change datastore objects even in suggestion_registry. 

I see this in suggestion_registry:
https://github.com/oppia/oppia/blob/4fe6c613334c4e3d178385dbf64996cd5b1fd543/core/domain/suggestion_registry.py#L349-L352

https://github.com/oppia/oppia/blob/4fe6c613334c4e3d178385dbf64996cd5b1fd543/core/domain/suggestion_registry.py#L463-L466

https://github.com/oppia/oppia/blob/4fe6c613334c4e3d178385dbf64996cd5b1fd543/core/domain/suggestion_registry.py#L599-L607

Reply:
Ah I see, I was looking for instances where the suggestion models were changed. I had put this in the registry file so that it becomes easy to generalize across many types (things like the accept, reject functions).

Reply:
So should I put the "assigning role to exploration"  & "sending emails" in the domain or the services?

Reply:
I think let's keep "sending emails" in the services. All the email related code was present in the services for feedback and suggestions.

For assigning roles, I am not very certain. We can probably keep it in the registry itself as each subclass will have a different entity type to assign to right? Maybe it might be possible to write a generic function for this in the service layer, if so that can be preferred.

Reply:
Okay, I'm leaving both of these (Sending email & assigning role) in the service layer as they are changing the models.

Also, I'll create an issue for refactoring suggestion_registry domain objects such that it doesn't change the models.

Reply:
Sg!